forge/src/executor/inspector/cheatcodes/mod.rs

```diff
@@ -0,0 +1,300 @@
+        // Record writes and reads if `record` has been called
+        if let Some(storage_accesses) = &mut self.accesses {
+            match interpreter.contract.code[interpreter.program_counter()] {
+                opcode::SLOAD => {
+                    let key = try_or_continue!(interpreter.stack().peek(0));
+                    storage_accesses
+                        .reads
+                        .entry(interpreter.contract().address)
+                        .or_insert_with(Vec::new)
+                        .push(key);
+                }
+                opcode::SSTORE => {
+                    let key = try_or_continue!(interpreter.stack().peek(0));
+                    storage_accesses
+                        .writes
+                        .entry(interpreter.contract().address)
+                        .or_insert_with(Vec::new)
+                        .push(key);
+                }
+                _ => (),
+            }
+        }
+
+        // Match logs if `expectEmit` has been called
+        if !self.expected_emits.is_empty() {
+            match interpreter.contract.code[interpreter.program_counter()] {
+                opcode::LOG0 => handle_expect_emit(self, interpreter, 0),
+                opcode::LOG1 => handle_expect_emit(self, interpreter, 1),
+                opcode::LOG2 => handle_expect_emit(self, interpreter, 2),
+                opcode::LOG3 => handle_expect_emit(self, interpreter, 3),
+                opcode::LOG4 => handle_expect_emit(self, interpreter, 4),
+                _ => (),
+            }
+        }
+
+        Return::Continue
+    }
+
+    fn call_end(
+        &mut self,
+        data: &mut EVMData<'_, DB>,
+        call: &CallInputs,
+        remaining_gas: Gas,
+        status: Return,
+        retdata: Bytes,
+        _: bool,
+    ) -> (Return, Gas, Bytes) {
+        if call.contract == *CHEATCODE_ADDRESS {
+            return (status, remaining_gas, retdata)
+        }
+
+        // Clean up pranks
+        if let Some(prank) = &self.prank {
+            data.env.tx.caller = prank.prank_origin;
+            if prank.single_call {
+                std::mem::take(&mut self.prank);
+            }
+        }
+
+        // Handle expected reverts
+        if let Some(expected_revert) = &self.expected_revert {
+            if data.subroutine.depth() <= expected_revert.depth {
+                let expected_revert = std::mem::take(&mut self.expected_revert).unwrap();
+                return match handle_expect_revert(false, &expected_revert.reason, status, retdata) {
+                    Err(retdata) => (Return::Revert, remaining_gas, retdata),
+                    Ok((_, retdata)) => (Return::Return, remaining_gas, retdata),
+                }
+            }
+        }
+
+        // Handle expected emits at current depth
+        if !self
+            .expected_emits
+            .iter()
+            .filter(|expected| expected.depth == data.subroutine.depth())
```

Comment:
i believe this should actually be a different comparison and is the reason for some of the expectEmit bugs

I have think on it more

Reply:
disregard wouldn't be here i don't think. should be in the actually emission of the logs